c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,522 @@
+                    // different rewrite.
+                    errs.push((span, rw, RewriteError::Conflict));
+                }
+                continue;
+            }
+
+            // Commit all rewrites that come strictly before `span`.  This leaves only nodes that
+            // at least partially overlap the current item.  These should normally be ancestors of
+            // the current item on the stack, but it's also possible that the current item
+            // erroneously partially overlaps a previous item.
+            while stack.last().map_or(false, |rt| rt.span.hi() <= lo) {
+                let rt = stack.pop().unwrap();
+                commit(&mut stack, &mut out, rt);
+            }
+
+            // Check for error cases.
+            let siblings = if let Some(parent) = stack.last() {
+                debug_assert!(parent.span.overlaps(span));
+                if !parent.span.contains(span) {
+                    // `span` overlaps `parent.span`, but isn't contained within it.
+                    errs.push((span, rw, RewriteError::PartialOverlap(parent.span)));
+                    continue;
+                }
+                &parent.children
+            } else {
+                &out
+            };
+
+            if let Some(prev) = siblings.last() {
+                debug_assert!(prev.span.lo() <= span.lo());
+                if prev.span.overlaps(span) {
+                    errs.push((span, rw, RewriteError::PartialOverlap(prev.span)));
+                    continue;
+                }
+            }
+
+            // TODO: check that all children are emitted somewhere in the `Rewrite`
+            // (e.g. given the rewrite `f(x + y) -> x`, we can't also have a rewrite on `x + y`,
+            // because that expression is discarded as part of the outer rewrite)
+            //
+            // Specifically: if there is a parent, check that either the current item's span is
+            // contained in the parent span and the parent's rewrite contains `Rewrite::Identity`,
+            // or the current item's span is contained in the span of some `Rewrite::Sub` in
+            // the parent rewrite.  If this doesn't hold, then produce `RewriteError::Discarded`.
+
+            // Push a new node onto the stack.
+            stack.push(RewriteTree {
+                span,
+                rw,
+                children: Vec::new(),
+            });
+        }
+
+        // There are no more children, so we can commit all remaining items on the stack.  Each
+        // item `stack[i+1]` is the last child of its parent `stack[i]`.
+        while let Some(rt) = stack.pop() {
+            commit(&mut stack, &mut out, rt);
+        }
+
+        (out, errs)
+    }
+}
+
+/// Split `rts` into the portion before `span`, the portion overlapping `span`, and the portion
+/// after `span`.  Nodes within `rts` should not overlap each other, and the list must be sorted by
+/// span; otherwise, the result are unspecified.
```

Comment:
i think itd be inexpensive and worthwhile to assert that it is in facto sorted.  maybe a debug assert